src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -373,6 +374,161 @@ void optimizeOrderBy(const ASTSelectQuery * select_query)
         elems = std::move(unique_elems);
 }
 
+/// Checks if ASTFunction or its arguments are stateful.
+bool isASTFunctionStateful(const ASTPtr & current_ast, const Context & context)
```

Comment:
The same. You need matcher for two types: ASTFunction and ASTExpressionList, Visitor based on it and `struct Data { bool has_stateful_function; };` as state